### ISSUE TITLE
Unable to view HTML exported notebook when offline

### ISSUE THREAD

**Issue by ForceBru** [marimo-team/marimo#575]
## How to reproduce

1. Create an empty notebook: `marimo edit random_notebook.py`. Fill it with random code like `print("Hello")`, execute & save.
2. Export the notebook as HTML.
3. Open the HTML file in the browser. I see the notebook, all fine.
4. Disconnect from the Internet, clear/disable browser caches, reload the HTML file. The point is to make sure nothing is loaded from caches.
5. All I see is a white page and errors like "couldn't load resource because no Internet connection" in the console. I suppose it's because the exported notebook can't load stuff like `<script type="module" crossorigin="anonymous" src="https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.1.77/dist/assets/index-z8usByTF.js">` since I'm offline and disabled caches.

## Why I think this is not good

IMO, exported notebooks should be _more_ accessible than the editable notebook that pops up with `marimo edit`. One should be able to read it almost anywhere, like a PDF document or an image. Especially given that the exported notebook is a _static_ HTML file with mostly text (my code) - I should be able to view it offline.

### Environment

{
  "marimo": "0.1.77",
  "OS": "Darwin",
  "OS Version": "22.6.0",
  "Processor": "i386",
  "Python Version": "3.12.1",
  "Binaries": {
    "Chrome": "--",
    "Node": "--"
  },
  "Requirements": {
    "black": "23.12.1",
    "click": "8.1.7",
    "jedi": "0.19.1",
    "pymdown-extensions": "10.7",
    "tomlkit": "0.12.3",
    "tornado": "6.4",
    "typing_extensions": "4.9.0"
  }
}

### Code to reproduce

Code doesn't matter, any code will do.

**Comment by mscolnick** [comment]
I totally see the use case. Couple reasons that could make this difficult (mostly thinking out loud)

1. Inlining the JS could get expensive and large. Right now we lazy-load some less-used plugins, so the single JS file you see there may actually end up loading many more (50+ JS files, some very small). It's not trivial to know which ones to include. 
2. We also load fonts from a CDN. These would not work offline either. Some fonts have fallbacks (basic text, markdown), but others don't like LaTex. It is not super trivial inlining these.

We have PNG export. Would a good PDF export satisfy your needs? Or would you like some interactivity still (e.g. hiding code, scrolling tables)?

**Comment by ForceBru** [comment]
> Would a good PDF export satisfy your needs?

Yep, I think having a _static_ export would be useful, especially for shareability, as stated in marimo's goals in the README. Just send someone a PDF and they can view it on a potato.

> Or would you like some interactivity still (e.g. hiding code, scrolling tables)?

Personally, I don't see the need for interactivity in an exported notebook. In an app - sure, in a notebook I'm currently running and editing - definitely. In a raw HTML file? Hiding code could be nice, maybe.

I think a simple (!) HTML file is already fine for static export:

1. HTML is ubiquitous and even washing machines probably have web browsers by now, so everyone can view it.
2. HTML is human-readable, so if someone doesn't have a browser for some reason, they still can read the markup, and thus the Python code (if it's included raw and not base64-encoded, like in Pluto).
3. HTML is structured, so in case GitHub goes up in flames or something, people would be able to parse their marimo notebooks and at least extract the code.
4. Images (and thus plots) can be included straight into the HTML `<img>` tag [using base64](https://base64.guru/developers/html/img). Also `<svg>`.
5. Of course, basic styles and interactivity can be added as a kind of "runtime" with simple CSS and JS.

As for inlining LaTeX, I found this just now: https://dev.to/uetchy/math-api-latex-math-as-svg-image-m4p. Apparently, it converts LaTeX markup to SVG images, so one can just put them into the HTML. It's [opensource](https://github.com/uetchy/math-api), so in theory, marimo could do something like this during export:

1. Given a LaTeX equation, convert it to an SVG with this.
2. Embed the SVG directly into the HTML.
3. Goto next equation.

Unfortunately, that's _just a theory_...

**Comment by mscolnick** [comment]
@jeffmelville, @ForceBru 
We have a way to HTML export now from the CLI: `marimo export html notebook.py -o notebook.html`. Could you use some other CLI tool (e.g. [pywebcopy](https://pypi.org/project/pywebcopy/, or plain `wget`) in order to make the html file offline?

**Comment by ForceBru** [comment]
__TL;DR__:

- This seems pretty complicated to me.
- Perhaps I could hack some code to convert a marimo notebook to a really basic HTML table, the main purpose of the resulting HTML being _simplicity_ and no requirement for JS or Internet connection.
- For that, having the ability to export a marimo notebook as detailed and structured JSON could be helpful. Say, `marimo export json` could produce a list of dictionaries representing cells. With that, people could come together and transpile that JSON to HTML, PDF and whatever else.

-------

> use some other CLI tool ... in order to make the html file offline

Not sure how that'd work? The [PyWebCopy documentation](https://pypi.org/project/pywebcopy/) says that "PyWebCopy can _"crawl" an entire website_ and download everything it sees in an effort to create a reasonable facsimile of the source website". Does it mean I should run PyWebCopy on the HTML generated by marimo? I guess that'd probably work, but why? This seems unnecessarily complicated. However, I likely don't know enough about webdev to understand what's complicated and what isn't. I really want to tinker with marimo and build a prototype HTML export tool that'd export to the simplest HTML possible that'd work without JavaScript on any potato.

Here's my basic plan if I ever have the time:

1. I assume marimo has a data structure that holds the full contents of the current notebook, so that the HTML generator can traverse it and extract cells and their contents.
1. Each marimo notebook becomes a table with one column.
2. Each marimo cell (a cell in that column) also becomes a table with one column and 3 rows:
   1. implied output (like a plot)
   2. code
   3. `print` output (just raw text)
3. Implied output is complicated because it can be very rich:
   - If it's a single image, base64-encode it and use `<img>` with [`src="data:..."`](https://stackoverflow.com/questions/8499633/how-to-display-base64-images-in-html). The goal is to make a _self-contained_ file, so images will have to become nasty base64 strings...
   - If it's a list or other structured data that can be collapsed, I'm thinking [`<details>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) might be of use.
   - If the structured output isn't massive (need to define "massive"), just put it in `<pre>` as raw text.
4. Code is colored text, so put each piece of syntax into a colored `<span>`. Make sure to not mess up indentation, keep comments and every minute detail of the code.
   - This can become complicated because this essentially needs to translate any Python AST to HTML and not lose any line positioning information.
   - This Python code becomes just text, not inside any `<textarea>` and not editable.
   - Always keep a copy of the raw code (as in the original `.py` file) in a collapsible (via `<details>`) `<textarea>` tag. The user should be able to easily extract the original code from the exported file.
6. Raw text output can be put in a `<textarea disabled>` tag. It's scrollable and thus doesn't take too much space when the output is very long.
7. Insert some vertical space between cells for aesthetics.

The result would be a really basic HTML table with almost no interactive elements (except for `<details>`):

- Minimal CSS (only used for code highlighting).
- No JS: it's an exported notebook and doesn't need to have anything dynamic, its main purpose is being _read, stored and shared_, not interacted with.
- Can be viewed by any software that can render HTML.

--------------

As usual, that's just a theory... I really should start hacking on some code to translate Python source to highlighted HTML - that shouldn't be too hard. If only marimo could export notebooks to well-structured JSON. In that case I could just ingest the JSON and convert it to HTML or whatever else, really.

**Comment by AdrienDart** [comment]
Hi!

I was wondering about this one too. Once implemented, will the (altair) charts be exported as static charts in the static html / pdf ?

Please do correct me if I'm wrong but any kind of export I'm trying to use leaves the charts empty. A static output of the chart within the export would be nice to have!

Thanks again for your time!

Adrien

**Comment by mscolnick** [comment]
@AdrienDart altair charts should render in the static, check out this example: https://static.marimo.app/charts-kman

**Comment by AdrienDart** [comment]
Thank you! Is there an alternative to publishing HTML to the web? Can we get the output as a local HTML saved locally? I'm working with private proprietary company data and I don't think I can use this static public app solution.
Thanks again for your help!

**Comment by mscolnick** [comment]
@AdrienDart, yes the Download options in the top right menu allow downloading the html (with or without the code)

**Comment by AdrienDart** [comment]
Thanks but when I use those exports, I get empty plots and empty dataframes (with the error: failed to fetch). Downloading the pdf of the app view, I get all the plots and dataframes but the output is like printing a webpage (not as nice). Apologies, I don't have a reproducible example for the moment as it is happening on my work computer.

**Comment by AdrienDart** [comment]
Hi, just FYI, plots and table are working fine in HTML when using plotly+mo.ui.plotly and mo.plain for the table

**Comment by n8henrie** [comment]
Similar here -- static html gives me `Error NetworkError when attempting to fetch resource.` for all tables. PDF chops all my charts in half on the page breaks. The WASM export seems interesting, but I'm planning to distribute the file over email to non-technical folks, so running a web server to load it is out of the question. It's not public data, so hosting on GH is out of the question.

Javascript console shows lots of:

```
TypeError: NetworkError when attempting to fetch resource.
    gG https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.10.17/dist/assets/index-UhR0gcZS.js:56
    http https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.10.17/dist/assets/index-UhR0gcZS.js:79
    hG https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.10.17/dist/assets/index-UhR0gcZS.js:56
index-UhR0gcZS.js:31:218156
```

but I am able to open `https://cdn.jsdelivr.net/npm/@marimo-team/frontend@0.10.17/dist/assets/index-UhR0gcZS.js` in my browser just fine.

**Comment by ozan-ozbeker-wvu** [comment]
> Hi, just FYI, plots and table are working fine in HTML when using plotly+mo.ui.plotly and mo.plain for the table

`mo.plain(df)` is exactly what I was looking for exporting the tables with *some* HTML formatting. 

Something similar to Quarto's `embed-resources` for HTML outputs would be great: <https://quarto.org/docs/reference/formats/html.html#rendering>. I'm not sure how different marimo's interactive tables in a session work from <https://machow.github.io/reactable-py/get-started/index.html>, but when you use `reactable-py` for tables in Quarto, the HTML output files will still have the interactive tables, so I believe it's possible at least. 

Currently, my Plotly charts remain interactive.

**Comment by chenyulue** [comment]
Any progress for this issue? OR any workaround for this? I want to share the exported html with other guys when offline. The exported pdf chops the figure and also has a bad layout.

**Comment by DGrady** [comment]
Just wanted to +1 this issue. The ability to `marimo export html` and get a document that avoids any attempted network requests would be incredibly useful. It seems like inlining all of marimo's dependencies should be possible; `marimo run` works just fine offline …

**Comment by ForceBru** [comment]
> incredibly useful

IMO this isn't just "incredibly useful", I think it's basic functionality. When I export a notebook, I want to receive an exact copy that's specifically made for _storage_ and _accessibility_. This in turn implies simplicity of the underlying representation: it doesn't need to look exactly like the actual marimo notebook, it could (should?) be barebones to allow _everyone_ to access it even on old potato PCs.

I think an exported notebook that requires Internet connection and **shows a blank page otherwise** simply defeats the purpose of exporting notebooks. Imagine you take a photo, it's saved on your device, it's right there, but you have to be connected to the Internet in order to view it. That's nonsense, right? Hopefully no billion-dollar company ever decides to implement this.

**Comment by AdrienDart** [comment]
> Any progress for this issue? OR any workaround for this? I want to share the exported html with other guys when offline. The exported pdf chops the figure and also has a bad layout.

For my use case, when using export html (no code included), plots using plotly are displayed, and tables using great-table are displayed.

**Comment by vascotenner** [comment]
I would like to be able to display these html exports also in 5 years from now. The marimo project could be stopped by then.
